### ISSUE TITLE
Allow multiple environment configurations in a single Chart

### ISSUE THREAD

**Issue by oforero** [helm/helm#6715]
At the moment Charts only read values from `values.yaml`, this makes it impossible to maintain different configurations per environment together with the Chart. The only alternative is to pass the environment-specific values from the command line. This mechanism works just fine until one attempts to use Umbrella charts, which forces you to blend the environment-dependent values for all the sub-charts into a single file.

This behaviour limits the utility of Umbrella charts by preventing their use as a simple aggregation of multiple services. In our use case we have different environments (e.g. dev1, dev2, sit, prod) and  the values files are archived as part of the Chart in a vars subdirectory, to install a chart we have to download the Chart, unpack it, and then run the `helm upgrade` using the unpacked directory. 

The process fails with Umbrella charts since there is no practical way to tell the template engine to look for a second file inside each of the sub-charts.

**Proposal:**

* Add an optional parameter `--environment <file-name>` to the relevant commands, like `helm template`
* The argument shouldn't be handled in the same way the `--values` arguments are handled, hence it won't be loaded right at the start, nor merged with all the other values (template.go:143)
* The argument would be pass to the `chartutil.Load` function, when the iterating over the files of the chart if the name of any file matches the parameter then it would be loaded and override `values.yaml`

**Example:**
`$ helm template my-chart --environment vars/dev/dev1.yaml`

The values would be loaded if the file exists, but it won't fail if it doesn't.

Allowing Umbrella charts to be used as a simple aggregation mechanism; all the other values overrides rules will still apply.

**Comment by hickeyma** [comment]
@oforero Have you tried tools like [helmfile](https://github.com/roboll/helmfile)?

**Comment by oforero** [comment]
I looked into it and it looks like a great tool but introduces some complexity which will make it hard to get implemented at my organization, not to mention the difficulty of getting a new tool approved and adopted for production use.

My understanding is that helmfile will require you to maintain the environment configuration separate to the chart. It will red the helmfile and the values files and pass them as parameters to helm. 

I will take another look, but I think there is real value to allowing a more flexible configuration natively, in some organizations is really complex to get new tools approved, especially for deployment to production environments.

**Comment by karuppiah7890** [comment]
@oforero I understand that it's hard to get a new tool approved and adopted. But bringing more and more features, will increase the complexity of helm, and will also increase venues for bugs. Currently it's pretty simple. Not sure if it's needed to build more features, especially when there are wrapper tools to help with deployment management for large scale deployments and for multiple environments. 

I'll let others put their opinion and I would request you to wait for others opinions before working more on the PR 😅 as if this proposal is not agreed, then the PR will be closed. And may be you can put that effort in helping helmfile to better that tool

**Comment by oforero** [comment]
@karuppiah7890 I understand, but my proposal does not increase Helm's complexity in any significant way. And it does simplify the life of users in a very concrete way because it enables people to keep the configuration for different environments together with their charts. 
I understand this is not that useful for tools for wide consumption, but for an internal project, it is desirable to keep all the configs for dev/sit/prod together with the chart. 
The lack of this feature forces us to write a lot of scripting to download unpack and upgrade the charts. And as I mentioned on this request it makes Umbrella Charts borderline unusable.

I have seen similar requests for something of this kind: https://github.com/helm/helm/issues/5076

I will take a look at helmfile again, but the issues I see with it will be harder to fix/address using helmfile:

* It forces you to keep the environment configurations separate from the charts. 
* It seems to use/depend on tiller, which I understood is being faced out, and that makes investing in its adoption a waste until they have migrated to helm3

As you can see my proposal makes minimal changes, it does not add infinite flexibility by allowing multiple overrides, it only allows one and reuses the same mechanism used by other values overrides. The key point is that it reads the file from inside the chart if exist, in the same way, it does with "values.yaml"

Please take a look again, IMHO this is a feature that belongs in Helm and it is easier to implement and maintain in helm, it might even simplify the code required in tools like helmfile or helmsman

**Comment by oforero** [comment]
BTW, I have invested a lot of time, but not necessarily with this, there is a problem with Helm3 not being able to read some index.yaml files, I think it might be due to a change in the yaml library. I hacked a "fix" but it is not a long term solution. I will write a ticket for it.

**Comment by karuppiah7890** [comment]
Actually, I'm also seeing some possible flaws (in my POV) in the design for the feature you are proposing. Let me tell what I feel -

First of all, charts having environment specific yaml files might not be a great idea. For example, there might be many teams, using the charts.
Each team will have environment specific configuration based on their use case. According to your solution, we would be putting a lot of environment
specific yamls in the chart, which doesn’t seem right - management becomes tougher. A lot of people would be changing the chart by putting or changing their environment specific yaml and would also be changing the version for each of these changes. Instead each team can manage their environment specific values yamls themselves, in a separate repo. 

Charts are like reusable functions with default values for the parameters they have. We can’t be putting all the parameters for all the possible different function calls in the same place as the function definition

About helmfile, I don't know about it or haven't used it, but I have seen some issues and PRs by @mumoshu in helmfile repo regarding helm 3 support 

https://github.com/roboll/helmfile/issues/668
https://github.com/roboll/helmfile/pull/842

**Comment by oforero** [comment]
@karuppiah7890 I see your point, I will look at this PRs and see if that might be the response to our issues.

I agree with you in the scenario you described putting the environment values outside the charts is a better approach, but I do not think my proposal is prescriptive and forces anyone to use it or move away from the approach they have at the moment. 

When I was working on this I try to achieve the following goals:

* It should not affect anyone not using the parameter. I think I achieved this.
* It should not introduce unbound flexibility that could work as a replacement for all the use cases the other `--values` parameters.

I disagree with you. Charts can be reusable functions, but they can also be lego bricks ready to be used adding any extra ingredients/parameters.

I think it is a question of different use cases, in our case teams maintain the code, the packaging, the environment parameters and the charts, in which case it makes a lot of sense to keep the environmental values in the chart and have version changes if the environment values change. Any configuration that can be modified without forcing a change in the version is exposed through a config map. Operations is a different team and hence there is real value in providing them with a fully self-contained chart that can be installed without having to add any steps/parameters, etc.

**Please note that the scenario you describe is still possible after my patch, this is an optional parameter and all the other functionality remains unaffected.**

**Comment by karuppiah7890** [comment]
I totally agree that existing use cases are still possible, but in my experience, I have seen that when there's more than one way to do something, it's confusing and unnecessary to the user. Especially first time users. 

I still am not sure how I would explain to a user that this feature is there and this is how it works etc, which is very very specific to your team's use case. I'm not sure if helm repo can accommodate everyone's specific use cases like this and build features and maintain them.

And yes, I agree with lego bricks. And even in my team, we use umbrella charts to deploy stateful components for different services. These charts have no template yamls, only requirements.yaml (with postgres, redis etc) and a default values.yaml overriding the values for the subcharts if needed, based on the service. And the environment specific values are kept separately. We also built a tool to manage multiple configs with ease, DRY, for ops teams. It's still not open sourced, but we feel charts and env specific values should be separate. And a full self contained chart should just be able to run in any environment with sensible defaults, and wouldn't need any extra steps. It will surely need extra parameters based on the user using it, and for the environment that's being used. For example chart will have a sane default for memory, but for prod, may be more is required. And that needn't be packed with the chart. I understand it makes it easy for your use case probably, but when adding the feature to helm, we need to consider a lot of things, in terms of design and existing tools and what's the responsibility of helm etc.

I'll stop there, as I've told enough. I'll let the core maintainers tell what they feel 😅

**Comment by oforero** [comment]
@karuppiah7890 I understand your point, and I agree that it can be confusing when there are many ways to archive the same results. I think in this case, the results are different, only similar. 

I would be happy to write a tutorial/example on how and when to use this approach over the other override options. When it comes to complexity, the move to remove tiller, it eliminates a lot of it and allowing users to avoid using other tools also removes complexity in specific scenarios. 

In the end, it is a trade-off, that it is the reason why I tried to keep the changes to a minimum and a very limited use case; and didn't duplicate the multiple overrides layers from outside the chart that helms support.

In our case, we decided to avoid defaults in most cases and have helm fail if variables are missing since in many cases we do not have sensible production defaults (E.g. URLs), and it is better to fail a deployment/templating than risking a bad deployment.

I look forward to the comments of the maintainers, thanks for your time and input.

**Comment by karuppiah7890** [comment]
cc @hickeyma @bacongobbler

**Comment by bacongobbler** [comment]
We're a little busy working on getting Helm 3 out the door right now... Holding off on looking at feature work for the time being. Thanks for being patient.

**Comment by oforero** [comment]
@bacongobbler I understand, I appreciate it.

**Comment by fritzb** [comment]
Multi environment value files is an important feature to have, and I think it should be part of helm core functionality. It is critical to store the value files in Helm repo seamlessly. +1 on this proposal.

**Comment by oforero** [comment]
> @bacongobbler @hickeyma Multi environment feature is an important feature to have, and I think it should be part of helm core functionality. It is critical to store the env files in Helm repo seamlessly. +1 on this proposal. Would be nice to propagate the environment all the way down to chart dependencies as well. Also can we have support of having to choose different value files which are stored in the helm repo ? (ie. helm install -f prod_values.yaml <-- where prod_values.yaml is stored in Helm repo)

@fritzb could you elaborate in what you would like to have?
The way I implemented it (https://github.com/helm/helm/pull/6742) the parameter value is treated as a file name and in the umbrella chart case, the parameter is passed to each of the requirements as a string, not the actual value files. Hence, each requirement chart will be configured with an independent environment values from a file inside the chart, if such file exists.

I have to test this in a more complex environment but unfortunately, our current artifactory has bad charts (versions like 1.0) and that makes the `helm repo add` fail.

**Comment by fritzb** [comment]
> @fritzb could you elaborate in what you would like to have?
> The way I implemented it (#6742) the parameter value is treated as a file name and in the umbrella chart case, the parameter is passed to each of the requirements as a string, not the actual value files. Hence, each requirement chart will be configured with an independent environment values from a file inside the chart, if such file exists.
> 
> I have to test this in a more complex environment but unfortunately, our current artifactory has bad charts (versions like 1.0) and that makes the `helm repo add` fail.

@oforero Probably something like:
1. value files gets loaded first
2. new directory called environment/ will host all the environment values which will override the default [1]. The name of the environment parameter could be the environment file name. Ie. europe_dev.yaml
3. pass the environment string to all dependencies charts and recurse.

All the above steps need to work even if the chart is stored in remote repository (ie. artifactory). I mentioned that -f <value file> to highlight that value files specified using -f will not pick up the value file stored inside the .tgz chart package downloaded from remote.

**Comment by hickeyma** [comment]
I am in agreement with @karuppiah7890 with regard to " environment specific configuration" in Helm core. I am still to see the overall benefit from general Helm users view point. Helmfile (when it updates to support Helm 3) provides the capability to satisfy different environment configuration while keeping the Helm configuration  interface simple and clear. 

@bacongobbler @mattfarina @thomastaylor312 it would be great to get your input as well.

**Comment by oforero** [comment]
@hickeyma depending on what you mean by general Helm users. Anyone who needs this flexibility without the extra complexity from using one more tool would benefit. I understand this seems very specific for in-house projects, but I suspect internal projects probably outnumber publicly available uses of Helm. Helm already supports Umbrella charts, this just enables a very simple use case for umbrella charts instead of requiring a third-party tool for it. Also, I think what those tools do is workaround this particular shortcoming, we have been trying to use Helmsman for this, and it involves hacks in python to concatenate different Yaml files into a single one, that is higher complexity and prone to breaking.

I am curious, why do you think the interface is made less clear by having this option? the default approach is still there and nobody will be forced to use this functionality.

**Comment by oforero** [comment]
@fritzb what I am proposing is not to pass another file that gets parsed and then the parsed values are passed to Helm. But to pass a file name to Helm, and Helm looking for that file inside the Chart and if it exists reading the file and overriding the values.yaml
The PR wrote for it works for umbrella charts, it passes the file name to each of the requirements if the sub chart does not contain a file with that name the parameter is ignored and only values.yaml is used. All the other overriding options still work as expected and are applied after loading the environment file.

**Comment by fritzb** [comment]
@oforero In my explanation it was the environment string/argument which will be passed recursively to the sub/dependency chart.

@hickeyma I agree with @oforero. There is no way to drive which value files to use in Umbrella's subchart. IMO this is a basic building block for Umbrella deployment in the real world.

**Comment by oforero** [comment]
Hello Guys,

Happy new year, any further thoughts on this? it is a feature that will really help us in the near/medium-term future at my current workplace.

regards,

Oscar

**Comment by mattfarina** [comment]
Jumping into this a little late.

I think this needs some serious investigation first in order to ferret out details. Helm is a package manager. Like apt, yum, etc. Going beyond that in scope would be bad for separation of concerns. There's a point where other tools (plural intentional) doing things in different styles should take over and use Helm. For example, Flux uses Helm in a GitOps model (pull based) while Helmfile updates using a push model.

**Comment by oforero** [comment]
@mattfarina I agree with you, but I do not see how this will move it beyond that scope. It will still be little more than APT, with only differences because of the nature of the platform it serves. It does not change at all.
The problem that my proposal & patch addresses are multiple environment configurations, at the moment you have to maintain the environment-specific values outside the charts. There are advantages to this approach, but also clear disadvantages in some situations like when teams are responsible for the deployments to all the environments and have to maintain the values required for them. At the moment we really have only two options:

* Maintain the environment values outside the charts
* Or, keep them in the charts and **creating scripts** to extract the files and pass them back to helm

This might be workable, but it is a waste of effort in my opinion. The bigger problem is that Umbrella charts become a whole lot less useful since you have to script to download and preprocess each of the component charts to extract the values files or maintain all environment values for all the component charts in a single values file. Which will become unmaintainable fast, in particular, if an application requires charts maintained by different teams.

We have tried Helmsman, and it also has this limitation, since it is intrinsic to Helm. To fix it in helmsman code would be required to download each chart, unpack it, get the environment file, and call helm in each chart with the right overrides. Helmfile and any other tools will independent fixes for it.

At the moment we have a Helmsman based hack that does something like this using python:

* Download & unpack each chart
* Concatenate all the environment files into a single one
* call helmsman with the concatenated file as a parameter

I oppose that solution because it is not robust, what happens if values have colliding names? or the indentation is different between files? etc.

Please take a look at the patch, you will notice is minimal, and will open this option to all of the tools if they decide to use it. Also, it is completely opt-in and charts that do not use it are not forced to deal with it or even know about it.

**Comment by oforero** [comment]
@Olefine has made a similar request in the past: https://github.com/helm/helm/issues/5076

Any tool will force you to chain your approach to the environment values and probably maintain a single big values overrides file, and such files require a lot of disciplined and careful maintenance, conflicting property names might occur if two teams use the same property name.

For example, if there is a requirement to have a different number of replicas per environment and different charts have a different number of replicas, the maintainer will be forced to use some sort of qualified name instead of simply `{{ .Values.replicas }}`.

**Comment by mattfarina** [comment]
I first want to note that this environment issue is a legitimate problem. I wish the options were better and that there were more/better competing options.

The question here is, does solving this problem fall within the scope of Helm (a package manager) or is it the scope of higher level tools? Scope creep is a real problem in software.

Helm as a package manager can be used in many ways. In push based systems (think ansible), in pull based systems (link Chef), as building blocks for other systems, and more. The building block part is the reason the package API was worked on so much for Helm v3.

Dealing with deployment environments has classically not been part of package managers. With Helm other tools have long served in this space.

A whole host of solutions can be created from shell scripts to much more complex tools (like ansible itself).

Why would this feature fall into scope for a package manager? I understand how it could work. The case needs to be made for it being in scope not that it could be done.

**Comment by oforero** [comment]
@mattfarina I understand, I thought that is what I've been trying to do. 

Let me try again:

The reason why I think it should be in Helm, and not in a tool higher in the stack, is that no other tool can implement this functionality in a clean and maintainable way and make it compatible with umbrella charts. 
If it is to be implemented by scripts or a tool like Helmsman extra steps would need to:
* read the dependencies of the umbrella charts
* download them 
* unpack them
* locate the relevant environment file
* create a single environment file from all environment files for each of the dependencies

The only other alternative is to have the configuration be maintained outside the charts themselves and probably in a single file with the configuration from different charts since there is no way to pass values to an umbrella chart that will be used to configure only one of the dependencies.

**Comment by bacongobbler** [comment]
would this use case be resolved if https://github.com/helm/helm/issues/4782 were implemented? I feel like tackling that issue would provide the same functionality provided here, and it may fit better as a solution that a package manager would provide, without introducing new terminology as "environments".

**Comment by fritzb** [comment]
See my posting on Nov 26, 2019-Dec 6, 2019. @bacongobbler This might work if the dependencies chart also doing the same thing (recursively), load the requested value file instead of 'values.yaml'. This is acceptable and more generic.

**Comment by oforero** [comment]
@bacongobbler I think the requests are very similar, I am not sure if that is requesting for having a single value file or two value files. If I understand correctly that request could be summarized by "check the chart internally for a file named like that before looking in the file system" If so this could work with one potential issue, umbrella charts. My proposal is that the umbrella chart simply passes the value of the parameter to its dependencies, and this might be problematic with an overloaded `--values` parameter. Another potential issue is that my proposal ignores the file if it doesn't exist instead of failing, which is the opposite of what the value flag does.

@fritzb I think loading only one file with a custom name will work but will force the repetition of configuration between environments. I think avoiding repetition is really, really important since it reduces the maintenance overhead.

**Comment by oforero** [comment]
@bacongobbler I have no issue with ditching the "environments" terminology, what I am not sure is if overloading the functionality of the "--values" flag would be a good idea, because of what it does in the failure scenarios and the behavior for dependencies. I also think it is important to allow the configuration to remain DRY.

For example the following scenario:

* I am deploying a chart, I would like to use a different value file inside the chart but the chart does not have the file and there is a local file with the same name. It might unintentionally deploy something different to the intention.

Would something like `--chart-values` be a better naming than environment values?
